stellargraph/utils/hyperbolic.py

```diff
@@ -0,0 +1,151 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright 2020 Data61, CSIRO
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+__all__ = [
+    "poincare_ball_distance",
+    "poincare_ball_exp",
+    "poincare_ball_mobius_add",
+]
+
+import tensorflow as tf
+import numpy as np
+
+
+# helper functions to manage numerical issues, inspired by https://github.com/dalab/hyperbolic_nn
+
+PROJECTION_EPS = 1e-5
+TANH_LIMIT = 15.0
+ATANH_LIMIT = tf.math.nextafter(1, 0)
+
+
+def _project(c, x):
+    """
+    Ensure ``x`` lies on the Poincaré ball with curvature ``-c``, in the presence of small numerical
+    errors.
+    """
+    max_norm = tf.math.rsqrt(c) * (1 - PROJECTION_EPS)
+    return tf.clip_by_norm(x, clip_norm=max_norm, axes=-1)
+
+
+def _tanh(x):
+    return tf.tanh(tf.clip_by_value(x, -TANH_LIMIT, TANH_LIMIT))
+
+
+def _atanh(x):
+    return tf.atanh(tf.clip_by_value(x, -ATANH_LIMIT, ATANH_LIMIT))
+
+
+def poincare_ball_mobius_add(c, x, y):
+    r"""
+    Möbius addition of ``x`` and ``y``, on the Poincaré ball with curvature ``-c``: :math:`\mathbf{x} \oplus^c \mathbf{y}`.
+
+    See Section 2 of [1] for more details.
+
+    [1] O.-E. Ganea, G. Bécigneul, and T. Hofmann, “Hyperbolic Neural Networks,” `arXiv:1805.09112 <http://arxiv.org/abs/1805.09112>`_, Jun. 2018.
+
+    Args:
+        c (tensorflow Tensor-like): the curvature of the hyperbolic space(s). Must be able to be
```

Comment:
Yeah, NumPy arrays. Basically anything that is supported by https://www.tensorflow.org/api_docs/python/tf/convert_to_tensor, because I think that's what functions like `tf.tanh` (etc.) support as input.